### PR TITLE
JAMES-2919 Validate JMAP partial read ADR after performance tests

### PR DESCRIPTION
**Pull request by chibenwa**
# Performance results

## Metadata

`Metadata read`: only mailboxIds + flags

 - GetMailboxes
 - GetMessageMist
 - 10x GetMessages 

With 5000 users per hour...

### Before:

Capture d’écran de 2019-12-04 17-38-10

### After:

Capture d’écran de 2019-12-04 17-38-14

However all the slow results (and few errors) occured upon projection population as this graphic shows:

Capture d’écran de 2019-12-04 17-38-17

A second run (on populated projection) means consistant, quick results:

Capture d’écran de 2019-12-04 17-38-20

Key metrics given 14 GetMessages request per second:

 - From 19% failure rate to 0
 - From 27.159 ms for succeding requests to 27 ms (x1000 inprovment of GetMessage command mean time)
 - At a global level, from 27.591 ms mean time (all requests) to 60 ms (x460 mean time improvment)

--------------------------------------------------

## Home loading

`FastView read`

 - GetMailboxes
 - GetMessageMist
 - 8x GetMessages 

2500 concurrent users

### Before:

Capture d’écran de 2019-12-05 09-47-37

### After:

Capture d’écran de 2019-12-05 09-47-39

Key metric:

 - From 7% errors to almost 0%
 - Mean time: From 12 802 ms to 407 ms (x14 improvment)
 - P99: From 60s (exceeding gatling time out at p95) to 1742 ms

### CODE REVIEW

File: src/adr/0012-jmap-partial-reads.md [hunk 1]
```diff
@@ -29,10 +27,21 @@ Some performance tests will be run in order to evaluate the improvements.
...
+ - Authenticate
+ - List mailboxes
+ - List messages in one of their mailboxes
+ - Get 10 time the properties expected to be fast to fetch with JMAP
+
+We went from:
+ - A 20% failure rate before this change to no failure
+ - Mean time for GetMessages went from 27 159 ms to 27 ms (1000 time improvment), for all operation from
+ 27 591 ms to 60 ms (460 time improvment)
+ - P99 is a metic thatdid not make sense because the initial simulation exceeded Gatling (the performance measuring tool
```
Comment (rouazana) [inline src/adr/0012-jmap-partial-reads.md]: metric that did

File: src/adr/0013-precompute-jmap-preview.md [hunk 2]
```diff
@@ -35,9 +33,20 @@ Some performance tests will be run in order to evaluate the improvements.
...
+ - Authenticate
+ - List mailboxes
+ - List messages in one of their mailboxes
+ - Get 10 time the properties expected to be fast to fetch with JMAP
+
+We went from:
+ - A 20% failure rate before this change to no failure
+ - Mean time for GetMessages went from 27 159 ms to 27 ms (1000 time improvment), for all operation from
+ 27 591 ms to 60 ms (460 time improvment)
+ - P99 is a metic thatdid not make sense because the initial simulation exceeded Gatling (the performance measuring tool
```
Comment (rouazana) [inline src/adr/0013-precompute-jmap-preview.md]: idem

File: src/adr/0013-precompute-jmap-preview.md [hunk 3]
```diff
@@ -35,9 +33,20 @@ Some performance tests will be run in order to evaluate the improvements.
...
+
+We went from:
+ - A 20% failure rate before this change to no failure
+ - Mean time for GetMessages went from 27 159 ms to 27 ms (1000 time improvment), for all operation from
+ 27 591 ms to 60 ms (460 time improvment)
+ - P99 is a metic thatdid not make sense because the initial simulation exceeded Gatling (the performance measuring tool 
+ we use) timeout (60s) at the p50 percentile. After this proposal p99 for the entire scenario is of 1 383 ms
 
-In case of a less than 5% improvement, the code will not be added to the codebase and the proposal will get the status 'rejected'.
+As such, this changeset significantly increase the JMAP performance.
```
Comment (remk) [inline src/adr/0013-precompute-jmap-preview.md]: s/increase/increases

File: src/adr/0012-jmap-partial-reads.md [hunk 4]
```diff
@@ -29,10 +27,21 @@ Some performance tests will be run in order to evaluate the improvements.
...
-
-In case of a less than 5% improvement, the code will not be added to the codebase and the proposal will get the status 'rejected'.
+GetMessages with a limited set of requested properties no longer result necessarily in full database message read. We
+thus have a significant improvement, for instance when only metadata are requested.
+
+Given the following scenario played by 5000 users per hour (constant rate)
+ - Authenticate
+ - List mailboxes
+ - List messages in one of their mailboxes
+ - Get 10 time the properties expected to be fast to fetch with JMAP
```
Comment (Arsnael) [inline src/adr/0012-jmap-partial-reads.md]: s/time/times

File: src/adr/0013-precompute-jmap-preview.md [hunk 5]
```diff
@@ -35,9 +33,20 @@ Some performance tests will be run in order to evaluate the improvements.
 
 ## Consequences
 
-We expect a huge performance enhancement for JMAP clients relying on preview for listing mails.
+Given the following scenario played by 5000 users per hour (constant rate)
+ - Authenticate
+ - List mailboxes
+ - List messages in one of their mailboxes
+ - Get 10 time the properties expected to be fast to fetch with JMAP
```
Comment (Arsnael) [inline src/adr/0013-precompute-jmap-preview.md]: idem

**Comment by chibenwa** [comment]
test this please

File: src/adr/0012-jmap-partial-reads.md [hunk 6]
```diff
@@ -29,10 +27,21 @@ Some performance tests will be run in order to evaluate the improvements.
...
+thus have a significant improvement, for instance when only metadata are requested.
+
+Given the following scenario played by 5000 users per hour (constant rate)
+ - Authenticate
+ - List mailboxes
+ - List messages in one of their mailboxes
+ - Get 10 times the mailboxIds and keywords of the given messages
+
+We went from:
+ - A 20% failure rate before this change to no failure
```
Comment (mbaechler) [inline src/adr/0012-jmap-partial-reads.md]: are they timeout or failures?

File: src/adr/0012-jmap-partial-reads.md [hunk 7]
```diff
@@ -29,10 +27,21 @@ Some performance tests will be run in order to evaluate the improvements.
...
+ - Mean time for GetMessages went from 27 159 ms to 27 ms (1000 time improvment), for all operation from
+ 27 591 ms to 60 ms (460 time improvment)
+ - P99 is a metric that did not make sense because the initial simulation exceeded Gatling (the performance measuring tool 
+ we use) timeout (60s) at the p50 percentile. After this proposal p99 for the entire scenario is of 1 383 ms
```
Comment (mbaechler) [inline src/adr/0012-jmap-partial-reads.md]: s/we use)/we use/

File: src/adr/0013-precompute-jmap-preview.md [hunk 8]
```diff
@@ -35,9 +33,20 @@ Some performance tests will be run in order to evaluate the improvements.
...
 
-We expect a huge performance enhancement for JMAP clients relying on preview for listing mails.
+Given the following scenario played by 2500 users per hour (constant rate)
+ - Authenticate
+ - List mailboxes
+ - List messages in one of their mailboxes
+ - Get 8 times the properties expected to be fast to fetch with JMAP
+
+We went from:
+ - A 7% failure rate before this change to almost no failure
```
Comment (mbaechler) [inline src/adr/0013-precompute-jmap-preview.md]: failures or timeout?

File: src/adr/0013-precompute-jmap-preview.md [hunk 9]
```diff
@@ -35,9 +33,20 @@ Some performance tests will be run in order to evaluate the improvements.
...
+ - Mean time for GetMessages went from 9 710 ms to 434 ms (22 time improvment), for all operation from
+ 12 802 ms to 407 ms (31 time improvment)
+ - P99 is a metric that did not make sense because the initial simulation exceeded Gatling (the performance measuring tool 
+ we use) timeout (60s) at the p95 percentile. After this proposal p99 for the entire scenario is of 1 747 ms
```
Comment (mbaechler) [inline src/adr/0013-precompute-jmap-preview.md]: same typo

File: src/adr/0013-precompute-jmap-preview.md (hunk 8 above)
Comment (chibenwa) [reply src/adr/0013-precompute-jmap-preview.md]: Both

File: src/adr/0012-jmap-partial-reads.md (hunk 6 above)
Comment (chibenwa) [reply src/adr/0012-jmap-partial-reads.md]: Both

File: src/adr/0012-jmap-partial-reads.md (hunk 7 above)
Comment (chibenwa) [reply src/adr/0012-jmap-partial-reads.md]: Nope it closes a bracket I opened before:

`Gatling (the performance measuring tool  we use)`

**Comment by chibenwa** [comment]
```
[963585cbc00fab0abcf8363041190cd69eb41d37] [ERROR] Failed to execute goal on project apache-james-backends-jpa: Could not resolve dependencies for project org.apache.james:apache-james-backends-jpa:jar:3.5.0-SNAPSHOT: Failed to collect dependencies at com.google.guava:guava:jar:25.1-jre -> org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Failed to read artifact descriptor for org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Could not transfer artifact org.codehaus.mojo:animal-sniffer-annotations:pom:1.14 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/org/codehaus/mojo/animal-sniffer-annotations/1.14/animal-sniffer-annotations-1.14.pom.part (No such file or directory) -> [Help 1]
```

test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**